pkg/natdiscovery/request_handle.go

```diff
@@ -0,0 +1,156 @@
+/*
+© 2021 Red Hat, Inc. and others.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natdiscovery
+
+import (
+	"net"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
+	proto2 "google.golang.org/protobuf/proto"
+	"k8s.io/klog"
+
+	"github.com/submariner-io/submariner/pkg/natdiscovery/proto"
+)
+
+func (nd *natDiscovery) handleRequestFromAddress(req *proto.SubmarinerNatDiscoveryRequest, addr *net.UDPAddr) error {
+	response := proto.SubmarinerNatDiscoveryResponse{
+		RequestNumber: req.RequestNumber,
+		Sender: &proto.EndpointDetails{
+			ClusterId:  nd.localEndpoint.Spec.ClusterID,
+			EndpointId: nd.localEndpoint.Spec.CableName,
+		},
+		Receiver: req.Sender,
+		ReceivedSrc: &proto.IPPortPair{
+			Port: uint32(addr.Port),
+			IP:   addr.IP.String(),
+		},
+	}
+
+	if req.Receiver == nil || req.Sender == nil || req.UsingDst == nil || req.UsingSrc == nil {
+		klog.Warningf("received NAT discovery packet %#v from %#v which seems to be malformed ", req, addr)
+
+		response.Response = proto.ResponseType_MALFORMED
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if req.Receiver.GetClusterId() != nd.localEndpoint.Spec.ClusterID {
```

Comment:
These are safety checks right? I mean, ideally we will not see these messages unless there is some bug in the code that framed the message and sent it.

Reply:
It could be if for example there is port mapping being done in a load balancer, or router, and they made a mistake between clusters when mapping the ports.